lockfile_unix.go

```diff
@@ -37,24 +39,40 @@ type lockfile struct {
 	lw       string
 	locktype int16
 	locked   bool
+	ro       bool
 }
 
-// Lock locks the lock file
-func (l *lockfile) Lock() {
+func (l *lockfile) lock(l_type int16, l_cmd int) {
 	lk := unix.Flock_t{
-		Type:   l.locktype,
+		Type:   l_type,
 		Whence: int16(os.SEEK_SET),
 		Start:  0,
 		Len:    0,
 		Pid:    int32(os.Getpid()),
 	}
-	l.mu.Lock()
-	l.locked = true
+	if !l.ro {
+		l.mu.Lock()
+		l.locked = true
+	}
 	for unix.FcntlFlock(l.fd, unix.F_SETLKW, &lk) != nil {
 		time.Sleep(10 * time.Millisecond)
 	}
 }
 
+// Lock locks the lock file as a writer
+func (l *lockfile) Lock() {
+	if l.ro {
+		l.RLock()
+	} else {
+		l.lock(unix.F_WRLCK, unix.F_SETLKW)
+	}
+}
+
+// LockRead locks the lock file as a reader
+func (l *lockfile) RLock() {
```

Comment:
Why not have a function RWLock()?

Reply:
The idea is to have `Lock()` for acquiring a write lock and `RLock()` for the read counter-part. It makes transitioning the code easier as we don't need to change all calls but only the ones where we access reading.